params/governance_paramset.go

```diff
@@ -0,0 +1,419 @@
+
+func init() {
+	for name, key := range govParamNames {
+		govParamNamesReverse[key] = name
+	}
+}
+
+// GovParamSet is an immutable set of governance parameters
+// with various convenience getters.
+type GovParamSet struct {
+	// Items in canonical type.
+	// Only type checked and validated values will be stored.
+	items map[int]interface{}
+}
+
+func NewGovParamSet() *GovParamSet {
+	return &GovParamSet{
+		items: make(map[int]interface{}),
+	}
+}
+
+func NewGovParamSetStrMap(items map[string]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for name, value := range items {
+		key, ok := govParamNames[name]
+		if !ok {
+			return nil, fmt.Errorf("Unknown governance param '%s'", name)
+		}
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetIntMap(items map[int]interface{}) (*GovParamSet, error) {
+	p := NewGovParamSet()
+
+	for key, value := range items {
+		err := p.set(key, value)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return p, nil
+}
+
+func NewGovParamSetChainConfig(config *ChainConfig) (*GovParamSet, error) {
+	items := make(map[int]interface{})
+	if config.Istanbul != nil {
+		items[Epoch] = config.Istanbul.Epoch
+		items[Policy] = config.Istanbul.ProposerPolicy
+		items[CommitteeSize] = config.Istanbul.SubGroupSize
+	}
+	items[UnitPrice] = config.UnitPrice
+	if config.Governance != nil {
+		items[GoverningNode] = config.Governance.GoverningNode
+		items[GovernanceMode] = config.Governance.GovernanceMode
+		if config.Governance.Reward != nil {
+			if config.Governance.Reward.MintingAmount != nil {
+				items[MintingAmount] = config.Governance.Reward.MintingAmount.String()
+			}
+			items[Ratio] = config.Governance.Reward.Ratio
+			items[UseGiniCoeff] = config.Governance.Reward.UseGiniCoeff
+			items[DeferredTxFee] = config.Governance.Reward.DeferredTxFee
+			items[StakeUpdateInterval] = config.Governance.Reward.StakingUpdateInterval
+			items[ProposerRefreshInterval] = config.Governance.Reward.ProposerUpdateInterval
+			if config.Governance.Reward.MinimumStake != nil {
+				items[MinimumStake] = config.Governance.Reward.MinimumStake.String()
+			}
+		}
+	}
+
+	return NewGovParamSetIntMap(items)
+}
+
+func (p *GovParamSet) set(key int, value interface{}) error {
+	ty, ok := govParamTypes[key]
+	if !ok {
+		return errors.New("Unknown governance param key")
+	}
+	parsed, ok := ty.ParseValue(value)
+	if !ok {
+		return errors.New("Malformed governance param value")
+	}
+	p.items[key] = parsed
+	return nil
+}
+
+func (p *GovParamSet) StrMap() map[string]interface{} {
+	m := map[string]interface{}{}
+	for key, value := range p.items {
+		m[govParamNamesReverse[key]] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) IntMap() map[int]interface{} {
+	m := map[int]interface{}{}
+	for key, value := range p.items {
+		m[key] = value
+	}
+	return m
+}
+
+func (p *GovParamSet) Get(key int) (interface{}, bool) {
+	v, ok := p.items[key]
+	return v, ok
```

Comment:
What about this?
```suggestion
	return p.items[key]
```

Reply:
That is not possible at least in go 1.18.

```
./governance_paramset.go:353:9: not enough return values
        have (interface{})
        want (interface{}, bool)
```